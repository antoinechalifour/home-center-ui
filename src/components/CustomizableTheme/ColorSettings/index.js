import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Button from 'ui/Button'
import ColorPicker from 'ui/ColorPicker'

export default function ColorSettings ({ theme, changeColor, commit, reset }) {
  const colorsSettings = theme.colors

  return (
    <Wrapper>
      {Object.keys(colorsSettings).map(setting => {
        const color = colorsSettings[setting]

        return (
          <Color key={setting}>
            <ColorPicker
              value={color}
              onChange={value => changeColor(setting, value)}
            />
            <span>{setting}</span>
          </Color>
        )
      })}

      <Actions>
        <Button onClick={reset}>Reset theme</Button>
        <Button primary onClick={commit}>Validate</Button>
      </Actions>
    </Wrapper>
  )
}

ColorSettings.propTypes = {
  theme: PropTypes.shape({
    colors: PropTypes.object.isRequired
  }).isRequired,
  changeColor: PropTypes.func.isRequired,
  commit: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired
}

const Wrapper = styled.div`
  background: #fff;
  position: relative;
  width: 100%;
  font-size: 75%;
  padding: 12px;

  display: flex;
  flex-direction: column;
  
  > div {
    padding: 12px;
  }
  
  @media (min-width: 860px) {
    flex-direction: row;
    align-items: center;
    justify-content: center;
  }
`

const Color = styled.div`
  display: flex;
  align-items: center;

  > :first-child {
    display: inline-flex;
    margin-right: 12px;
  }
`

const Actions = styled.div`
  display: flex;
  flex-direction: column;

  button {
    flex: 1;
  }

  button + button {
    margin-top: 4px;
  }
`
