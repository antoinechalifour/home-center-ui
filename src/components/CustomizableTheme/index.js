import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled, { ThemeProvider } from 'styled-components'
import deepMerge from 'deepmerge'
import ColorsIcon from 'react-icons/lib/md/invert-colors-on'
import ColorSettings from './ColorSettings'

const LOCAL_STORAGE_KEY = 'theme:colors'
const emptyThemeState = {
  colors: {},
  font: {}
}
let defaultThemeState = emptyThemeState

try {
  const themeStr = window.localStorage.getItem(LOCAL_STORAGE_KEY)

  if (themeStr) {
    defaultThemeState = JSON.parse(themeStr)
  }
} catch (err) {
  // Ignore
}

export default class CustomizableTheme extends Component {
  static propTypes = {
    baseTheme: PropTypes.shape({
      colors: PropTypes.shape({
        background: PropTypes.string.isRequired,
        backgroundInverse: PropTypes.string.isRequired,
        text: PropTypes.string.isRequired,
        textInverse: PropTypes.string.isRequired,
        primary: PropTypes.string.isRequired,
        accent: PropTypes.string.isRequired,
        light: PropTypes.string.isRequired
      }).isRequired,
      font: PropTypes.shape({
        size: PropTypes.string.isRequired,
        family: PropTypes.string.isRequired,
        lineHeight: PropTypes.string.isRequired
      }).isRequired
    }).isRequired
  }

  state = {
    theme: defaultThemeState,
    showSettings: false
  }

  _toggleSettings = () =>
    this.setState({
      showSettings: !this.state.showSettings
    })

  _onColorChange = (setting, color) =>
    this.setState(lastState => {
      return {
        theme: {
          ...lastState.theme,
          colors: {
            ...lastState.theme.colors,
            [setting]: color
          }
        }
      }
    })

  _onReset = () =>
    this.setState({ theme: emptyThemeState, showSettings: false })

  _onCommit = () => {
    window.localStorage.setItem(
      LOCAL_STORAGE_KEY,
      JSON.stringify(this.state.theme)
    )
    this.setState({ showSettings: false })
  }

  render () {
    const theme = deepMerge(this.props.baseTheme, this.state.theme)

    return (
      <ThemeProvider theme={theme}>
        <Container>
          {this.props.children}

          <IconContainer>
            <ColorsIcon onClick={this._toggleSettings} />
          </IconContainer>

          <Modal isVisible={this.state.showSettings}>
            <ColorSettings
              theme={theme}
              changeColor={this._onColorChange}
              reset={this._onReset}
              commit={this._onCommit}
            />
          </Modal>
        </Container>
      </ThemeProvider>
    )
  }
}

const Container = styled.div`
  font-family: ${({ theme }) => theme.font.family};
  font-size: ${({ theme }) => theme.font.size};
  line-height: ${({ theme }) => theme.font.lineHeight};
  background: ${({ theme }) => theme.colors.background};
  color: ${({ theme }) => theme.colors.text};
  letter-spacing: 0.04rem;
`

const IconContainer = styled.div`
  position: fixed;
  left: 0;
  bottom: 0;
  z-index: 10;
  color: inherit;
  font-size: 16px;
  padding: 2px 12px;
  background: rgba(255, 255, 255, .65);
  color: #373d3f;
  border-top-right-radius: 4px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, .24);
`

const Modal = styled.div`
  position: fixed;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 20;
  display: flex;
  align-items: center;
  justify-content: center;
  transition: transform .2s ease-in;
  transform: ${({ isVisible }) => (isVisible ? 'translateY(0)' : 'translateY(100%)')};
`
