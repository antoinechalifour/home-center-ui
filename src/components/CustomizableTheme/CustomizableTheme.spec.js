import React from 'react'
import { shallow } from 'enzyme'
import CustomizableTheme from '.'

const baseTheme = {
  colors: {
    background: '#fff',
    backgroundInverse: 'red',
    text: 'blue',
    textInverse: 'green',
    primary: 'yellow',
    accent: 'cyan',
    light: 'orange'
  },
  font: {
    size: '18px',
    family: 'Nunito',
    lineHeight: '1.6'
  }
}

describe('CustomizableTheme', () => {
  it('Should render correctly (settings are hidden)', () => {
    const props = { baseTheme }
    const wrapper = shallow(
      <CustomizableTheme {...props}>
        <div>here are children</div>
      </CustomizableTheme>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('Should render correctly (settings are visible)', () => {
    const props = { baseTheme }
    const wrapper = shallow(
      <CustomizableTheme {...props}>
        <div>here are children</div>
      </CustomizableTheme>
    )

    wrapper.setState({ showSettings: true })

    expect(wrapper).toMatchSnapshot()
  })

  describe('_toggleSettings()', () => {
    it('Should set the correct state', () => {
      const props = { baseTheme }
      const wrapper = shallow(
        <CustomizableTheme {...props}>
          <div>here are children</div>
        </CustomizableTheme>
      )

      wrapper.instance()._toggleSettings()

      expect(wrapper.instance().state).toEqual({
        showSettings: true,
        theme: {
          colors: {},
          font: {}
        }
      })
    })
  })

  describe('_onColorChange()', () => {
    it('Should set the correct state', () => {
      const props = { baseTheme }
      const wrapper = shallow(
        <CustomizableTheme {...props}>
          <div>here are children</div>
        </CustomizableTheme>
      )

      wrapper.instance()._onColorChange('background', 'cyan')

      expect(wrapper.instance().state).toEqual({
        showSettings: false,
        theme: {
          colors: {
            background: 'cyan'
          },
          font: {}
        }
      })
    })
  })

  describe('_onReset()', () => {
    it('Should set the correct state', () => {
      const props = { baseTheme }
      const wrapper = shallow(
        <CustomizableTheme {...props}>
          <div>here are children</div>
        </CustomizableTheme>
      )

      wrapper.setState({
        showSettings: true,
        theme: {
          colors: {
            background: 'cyan'
          },
          font: {}
        }
      })
      wrapper.instance()._onReset()

      expect(wrapper.instance().state).toEqual({
        showSettings: false,
        theme: {
          colors: {},
          font: {}
        }
      })
    })
  })

  describe('_onCommit()', () => {
    beforeEach(() => {
      global.localStorage.setItem.mockReset()
    })

    it('Should store the current theme in the local storage and set the correct state', () => {
      const props = { baseTheme }
      const wrapper = shallow(
        <CustomizableTheme {...props}>
          <div>here are children</div>
        </CustomizableTheme>
      )

      wrapper.setState({
        showSettings: true,
        theme: {
          colors: {
            background: 'cyan'
          },
          font: {}
        }
      })
      wrapper.instance()._onCommit()

      expect(window.localStorage.setItem.mock.calls[0][1]).toEqual(
        '{"colors":{"background":"cyan"},"font":{}}'
      )

      expect(wrapper.instance().state).toEqual({
        showSettings: false,
        theme: {
          colors: {
            background: 'cyan'
          },
          font: {}
        }
      })
    })
  })
})
