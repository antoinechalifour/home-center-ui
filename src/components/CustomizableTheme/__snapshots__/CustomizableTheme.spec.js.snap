// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomizableTheme Should render correctly (settings are hidden) 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "accent": "cyan",
        "background": "#fff",
        "backgroundInverse": "red",
        "light": "orange",
        "primary": "yellow",
        "text": "blue",
        "textInverse": "green",
      },
      "font": Object {
        "family": "Nunito",
        "lineHeight": "1.6",
        "size": "18px",
      },
    }
  }
>
  <styled.div>
    <div>
      here are children
    </div>
    <styled.div>
      <MdInvertColorsOn
        onClick={[Function]}
      />
    </styled.div>
    <styled.div
      isVisible={false}
    >
      <ColorSettings
        changeColor={[Function]}
        commit={[Function]}
        reset={[Function]}
        theme={
          Object {
            "colors": Object {
              "accent": "cyan",
              "background": "#fff",
              "backgroundInverse": "red",
              "light": "orange",
              "primary": "yellow",
              "text": "blue",
              "textInverse": "green",
            },
            "font": Object {
              "family": "Nunito",
              "lineHeight": "1.6",
              "size": "18px",
            },
          }
        }
      />
    </styled.div>
  </styled.div>
</ThemeProvider>
`;

exports[`CustomizableTheme Should render correctly (settings are visible) 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "accent": "cyan",
        "background": "#fff",
        "backgroundInverse": "red",
        "light": "orange",
        "primary": "yellow",
        "text": "blue",
        "textInverse": "green",
      },
      "font": Object {
        "family": "Nunito",
        "lineHeight": "1.6",
        "size": "18px",
      },
    }
  }
>
  <styled.div>
    <div>
      here are children
    </div>
    <styled.div>
      <MdInvertColorsOn
        onClick={[Function]}
      />
    </styled.div>
    <styled.div
      isVisible={true}
    >
      <ColorSettings
        changeColor={[Function]}
        commit={[Function]}
        reset={[Function]}
        theme={
          Object {
            "colors": Object {
              "accent": "cyan",
              "background": "#fff",
              "backgroundInverse": "red",
              "light": "orange",
              "primary": "yellow",
              "text": "blue",
              "textInverse": "green",
            },
            "font": Object {
              "family": "Nunito",
              "lineHeight": "1.6",
              "size": "18px",
            },
          }
        }
      />
    </styled.div>
  </styled.div>
</ThemeProvider>
`;
